include/vapor/ArbitrarilyOrientedRegularGrid.h

```diff
@@ -0,0 +1,103 @@
+#pragma once
+
+#include <ostream>
+#include <vector>
+#include <vapor/common.h>
+#include <vapor/RegularGrid.h>
+#include <glm/glm.hpp>
+
+// clang-format off
+
+struct planeDescription {
+    size_t sideSize;
+    std::vector<double> origin;
+    std::vector<double> rotation;
+    VAPoR::CoordType boxMin;
+    VAPoR::CoordType boxMax;
+    VAPoR::CoordType domainMin;
+    VAPoR::CoordType domainMax;
+};
+
+namespace VAPoR {
+
+//! Create a 2D RegularGrid that is rotated along an arbitrary orientation.  This new grid
+//! samples a given 3D Grid object along a plane, through the entire domain.  The "planeDescription" 
+//! variable includes parameters that define the 2D plane's origin, rotation, and valid extents.  Points on the
+//! 2D RegularGrid that are outside of the valid extents will be assigned as
+//! missingValues.
+//!
+//! Note: The boxMin and boxMax values within the planeDescription should come from
```

Comment:
The need to pass in both the box and domain extents is really confusing to me. It should not be necessary. grid3d knows its extents (Grid::GetUserExtents). box{Min,Max} constrain the region of the plane that you wish to intersect with grid3d. The box may lie entirely inside, outside, or intersect grid3d's extents. The code should handle all possibilities. So I guess a question is why are "erroneous results" produced when grid3d's extents used? Sounds like a bug that we should fix, rather than complicate the API and not have the code be robust.

Reply:
This was done last week in PR #2972, to fix #2971.  We had a bit of a conversation on it and everyone gave a thumbs up before the merge.  I'll copy/paste my response from last week below.  The PR linked above will have more context.  

Let me know if there's still something that's unclear though.

> The SliceRenderer was crashing when the Box and plane did not intersect. This was happening because the returned 2D grid was only being sampled within the bounds of the current Box. When the plane and Box did not intersect the returned grid would be invalid, causing a crash.

> One option was to add logic in the SliceRenderer to handle this special case, however this option would require every renderer using SliceGridAlongPlane() to handle it as well.

> The alternative I chose was to have SliceGridAlongPlane() return a valid grid in this case. The returned grid would be sampled along a given plane, and would have missingValues at points that do not lie within the Box. The Slice and Contour logic would be simplified by letting them operate on the returned grid in the same way, whether there are valid values or not.

> Note that I am using the domain extents gathered from the DataMgr, and the Box extents from RenderParams. I'm not using the extents gathered from the 3D grid because those can be in disagreement with the DataMgr and RenderParams. This occurs due to roundoff whenever the Grid's extents are reduced, which incrementally reduces the Grid's dimension lengths.

> Hope that makes sense.

Reply:
Users of this class need to be able to handle the case where the grid consists entirely of missing values. That is possible even if the plan is located   entirely within the valid domain. So I don't understand the argument that "When the plane and Box did not intersect the returned grid would be invalid, causing a crash.". It should not crash simply because the plane is outside of the grid and contains nothing but missing values.

Reply:
> It should not crash simply because the plane is outside of the grid and contains nothing but missing values.

It does not crash in this case.  Let me see if this helps:

Previous to #2971, the extents of the plane were defined according to the Box.  I would find where the plane intercepts the edges of the Box, and use these intercepts to find the Minimum Area Rectangle (MAR) enclosing all of these intercepts.

When the plane did not intersect the Box, there would be no intercepts and the MAR algorithm would crash.  One option was to return a nullptr in this case.  However this would 1) require logic in the Slicer to handle this case, and 2) require each Renderer to handle the special case where the returned grid is invalid.

Instead, we can find where the plane intercepts the edges of the domain, because the plane will always lie within the domain.  When we sample along the plane inside the domain, we can simply attribute missing values to coordinates outside of the Box.  Coordinates inside the box can still attribute MissingValues.

This way the Renderer can always rely on getting a valid grid according to the parameters its using.  No special case handling is needed.  If the Renderer gets a slice that has all missing values it operates on it the same as if there were scalar data.

Reply:
@clyne - I've changed the logic to use the Grid's extents instead of the full domain extents and it's passed my tests.

Reply:
Cool beans. The Doxygen documentation for the class needs to be refreshed for this change (it still discusses domainMin, domainMax).

Reply:
I missed that one, and an additional reference above.  Thanks for the catch.

Reply:
Note: the (outdated) label is missing here too.  Now sure why.  Current changes can be seen [here](https://github.com/NCAR/VAPOR/pull/2974/commits/7bb2296c47ce030971e08910ed522cfb931c78aa).